### ISSUE TITLE
Very long duration in 'parsing includes' every time open a file, without any modification

### ISSUE THREAD

**Issue by lixin-wei** [clangd/clangd#1165]
I have background indexing enabled, indexing completed, disk cache saved.

But when I open a file, it needs 10+ seconds before the definition jump gets ready. It's not the first time, when I close this file and open this file again, it still needs 10+ seconds.

**System information**

My project is very large. For small files with fewer includes, the speed is higher(1s). But for large files with a lot of includes, it will take 10+ seconds.

Output of `clangd --version`:
```
clangd version 14.0.3 (https://github.com/llvm/llvm-project 1f9140064dfbfb0bbda8e51306ea51080b2f7aac)
Features: linux+grpc
Platform: x86_64-unknown-linux-gnu
```
Editor/LSP plugin:
VSCode 1.67.2  with clangd plugin v0.1.17

Operating system: MacOS 12.3.1 (21E258)

**Logs**
In this log I opened the same file 3 times. Every time it takes 10+ seconds to get ready.
[clangd.log](https://github.com/clangd/clangd/files/8771035/clangd.log)

**Comment by HighCommander4** [comment]
> But when I open a file, it needs 10+ seconds before the definition jumping gets ready.

This is a known limitation of clangd's current design. Clangd needs to build an AST for a file when opening it, which requires parsing the file and its transitive includes, and it can take this long.

The index augments the AST to enable queries that span multiple translation units such as find-references, but does not replace the need for the AST.

Here's an assorted list of ideas and works in progress for improving the current situation:

 * https://discourse.llvm.org/t/rfc-a-c-pseudo-parser-for-tooling/59217
 * https://reviews.llvm.org/D112661
 * https://github.com/clangd/clangd/issues/391
 * https://github.com/clangd/clangd/issues/589

> When I close this file and open this file again, without any modification, it still needs 10+ seconds.

The is also known and appears to be intentional, as discussed in https://github.com/clangd/clangd/issues/312.

**Comment by lixin-wei** [comment]
@HighCommander4 Thanks for the clarification! But I'm still wondering why we can't cache the AST in disk? So we don't need to parse it again.

**Comment by HighCommander4** [comment]
> But I'm still wondering why we can't cache the AST in disk?

We do some caching of ASTs (more specifically of _preambles_, which is the part of the AST that's most expensive to build), based on issues like #255 and #1007 (which concern the storage location for this cache).

I'm not clear on what the lifetime of the entries in this cache is, and whether it might be suitable for fixing #312.

However, I think I can answer the question of why we don't cache them **for the whole project** (e.g. as a by-product of indexing): the disk space usage would be prohibitive. Glancing at the preamble files stored in my temp directory, I'm seeing sizes on the order of 100 MB. For a project with, say, 8000 source files, preambles for every file would take up 800 GB of space!

**Comment by lixin-wei** [comment]
@HighCommander4 I see. I think we should provide an option for this cache size instead of expiring it immediately after closing. That will make more sense.

**Comment by HighCommander4** [comment]
> We do some caching of ASTs (more specifically of _preambles_, which is the part of the AST that's most expensive to build), based on issues like #255 and #1007 (which concern the storage location for this cache).
> 
> I'm not clear on what the lifetime of the entries in this cache is, and whether it might be suitable for fixing #312.

I did some code reading, and I think I was mistaken to call this storage a "cache". It looks like the current intention is just to store the preamble on disk while the file is open, to avoid the memory usage cost of keeping it in memory. The preamble file is intended to be deleted when the file is closed; the reason I saw some files left over is probably because clangd crashed on some previous occasions and the cleanup didn't happen.

Turning this into a cache would require adding some new functionality, such as determining whether a previously built preamble is "up to date" and can be used for a new open of a file (presumably, based on something like a hash of the source file's preprocessed contents?).

I do wonder if such a cache would be useful in practice -- it's very possible that, depending on the cache size, it would either fill up too quickly to be of much help, or consume more disk space than most users would find reasonable.

If someone's interested in experimenting with this, I'm happy to provide guidance.

**Comment by lixin-wei** [comment]
> I do wonder if such a cache would be useful in practice -- it's very possible that, depending on the cache size, it would either fill up too quickly to be of much help, or consume more disk space than most users would find reasonable.

@HighCommander4 I don't think so. Actually in practice, we'll only focus on a small subset of files when developing(maybe ~50 files). So the requirement of the cache size won't be too large.

> If someone's interested in experimenting with this, I'm happy to provide guidance.

I'm very interested in this. Could you please help me?

**Comment by lixin-wei** [comment]
@lightmelodies Awesome! What you did is exactly the 'AST cache' that I'm talking about with @HighCommander4, you just didn't implement functionalities like determining whether a previously built preamble is "up to date" after clangd is restarted. Besides, you won't expire them forever.

@HighCommander4 Maybe we can implement an in-memory cache first as @lightmelodies did?

**Comment by HighCommander4** [comment]
@lightmelodies Thanks for the very interesting pointers! The cache you are referring to is one I didn't know about. It's at a higher level and caches entire ASTs (preamble + main file), with the main-file portion stored in memory.

As you observed, currently a file's ASTs is evicted from this cache when the file is closed. So what is the purpose of the cache? I didn't realize this before, but it appears that **clangd doesn't even retain the ASTs of all open files in memory**, only the most recently accessed 3. However, it retains the **preambles** of every open file.

So, if you open 4 files in your editor, and you cycle through them, the main-file portions of their ASTs are actually rebuilt every time!

This design choice seems a bit odd to me, but I guess it has the potential to keep clangd's memory usage low (since only preambles are retained for every open file, and those are stored on disk), and the performance impact is fairly small (since the main-file part of the AST is usually much faster to build than the preamble).

**Comment by lixin-wei** [comment]
@HighCommander4 So in conclusion, I think it is worth it to implement this cache. What do you think? Could you please give me some guidance on how to experiment with it? I'm suffering from such low speed.

**Comment by HighCommander4** [comment]
@lixin-wei Here are my thoughts on an implementation approach suitable for experimentation:

First, I do not think that extending the `ASTCache` which @lightmelodies modified is suitable for this purpose. The keys in this cache are pointers to `ASTWorker` objects, which store per-file state for open files. The correct operation of this cache relies on the file being open, not just in the sense of the `ASTWorker` object existing (which @lightmelodies handled by keeping it around forever), but also on the file being open in the LSP client and the client sending clangd a notification every time the file is modified -- this will definitely not be the case for files which are closed.

Instead, I think it makes sense to add a **new cache** for the purpose of retaining the ASTs of closed files.

Further, I think it would make sense for this cache to retain **preambles only**. (Otherwise, it would interact strangely with the existing `ASTCache`, where we would retain the main-file portions of the AST of previously closed files, but not necessarily of all open files...) Retaining the preambles should give us the performance gain we want, since that's the expensive part of AST to build.

Sketch of an implementation:

 * The cache can live in `TUScheduler`, like the [existing ASTCache](https://searchfox.org/llvm/rev/a13b61f7f0a2a44546fb39663b0d6f2547841e02/clang-tools-extra/clangd/TUScheduler.cpp#650).
 * The cache keys can be file paths.
 * The cache value can be a structure that includes:
     * The preamble itself (`shared_ptr<PreambleData>` or similar)
     * Some information that allows us to determine whether the preamble is up to date. The details here need to be worked out; the function [`isPreambleCompatible`](https://searchfox.org/llvm/rev/a13b61f7f0a2a44546fb39663b0d6f2547841e02/clang-tools-extra/clangd/Preamble.h#110) may be relevant.
 * When an `ASTWorker` is [destroyed](https://searchfox.org/llvm/rev/a13b61f7f0a2a44546fb39663b0d6f2547841e02/clang-tools-extra/clangd/TUScheduler.cpp#788) (happens on file-close), its preamble (stored in `ASTWorker::LatestPreamble` I believe) should be stored in the cache.
 * When an `ASTWorker` is [created](https://searchfox.org/llvm/rev/a13b61f7f0a2a44546fb39663b0d6f2547841e02/clang-tools-extra/clangd/TUScheduler.cpp#751) (happens on file-open), we need to:
     * Look up the file path in the cache
     * If there is an entry in the cache, check if the preamble is up to date, and if so use it instead of building a new one
 * We'll want to give the cache a size limit, and respect it by evicting the oldest entry when it gets full.

Hopefully that gives you a starting point. There are a fair amount of details to be worked out, please feel free to ask questions here or in the clangd Discord channel.

**Comment by HighCommander4** [comment]
It may also be helpful to look at https://reviews.llvm.org/D112661, which also adds a cache of preambles of closed files, though for a slightly different / more ambitious purpose (reusing them in a heuristic way when opening **other** files with a similar preamble).

**Comment by lightmelodies** [comment]
@HighCommander4 We can retain cache for preambles only. But instead of complex reuse checking, the lsp client will send notify when file is modified, which should keep preambles update to date in most cases.  From my experience, comment ``WorkScheduler->remove(File)`` (which keep AST woker forever) without any other modification already work well.

**Comment by HighCommander4** [comment]
> the lsp client will send notify when file is modified

Not for closed files, right? Keep in mind that files can be modified outside of the editor, e.g. via `git pull` (or even just in another editor).

**Comment by morethanlessoo7** [comment]
Hi, I encountered this problem too. is this problem solved in recent release?
